### PR TITLE
[WIP] Truncate labels for k8s

### PR DESCRIPTION
**Pull request by pinin4fjords**
This PR proposes automated truncation of labels for k8s where they exceed the 63 character limit, see https://github.com/nextflow-io/nextflow/issues/1577.

### CODE REVIEW

File: modules/nextflow/src/main/groovy/nextflow/k8s/model/PodSpecBuilder.groovy [hunk 1]
```diff
@@ -236,10 +239,24 @@ class PodSpecBuilder {
...
+                }
+                else if( entry.value.size() > 63 ) {
+                    log.debug "K8s label exceeds allowed size: 63 -- offending entry=$entry"  
+                    labels.put(entry.key, entry.value.substring(0,63))   
+                }
+                else {
+                    labels.put(entry.key, entry.value)    
+                }
+            } 
         }
```
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/k8s/model/PodSpecBuilder.groovy:260]: Make sense to separate this snippet into its own method other this method too complex, something like 

```
Map<String, String> normalizeLabels(Map<String, String> labels) {
 // the same as above
}
```

The it becomes:

```
if( opts.labels ) {
  labels.putAll(normaliseLabels(opts.label)) 
}

File: modules/nextflow/src/test/groovy/nextflow/k8s/model/PodSpecBuilderTest.groovy [hunk 2]
```diff
@@ -103,6 +103,52 @@ class PodSpecBuilderTest extends Specification {
...
+                           containers:[
+                                   [name:'foo',
+                                    image:'busybox',
+                                    command: ['sh', '-c', 'echo hello'],
+                                    workingDir:'/some/work/dir'
+                                   ]
+                           ]
+                   ]
+        ]
+    }
```
Comment (pditommaso) [inline modules/nextflow/src/test/groovy/nextflow/k8s/model/PodSpecBuilderTest.groovy:152]: This can become a test only for `normalizeLabels`.

**Comment by pinin4fjords** [comment]
Thanks @pditommaso, will do. 

Note that it doesn't actually work yet anyway(hence the WIP), was just running it past the CI to be sure. For some reason the labels still come out too long (including with my original real world test in the linked issue). Suggestions as to why most welcome!

**Comment by pditommaso** [comment]
Didn't check the test results, but surely a good practice is to run tests locally before committing using the cmd `make smoke`.

**Comment by pinin4fjords** [comment]
Apologies, just wanted to make sure I wasn't doing something wrong locally- seems like this should work. Also I hadn't come across 'make smoke' in the docs.

**Comment by pditommaso** [comment]
`make smoke` is just a quicker version of `make test`

**Comment by pinin4fjords** [comment]
Apologies- this now passing tests so I pushed, however, real world use case still appears to have long tags, will add further commits once I've addressed that.

**Comment by pinin4fjords** [comment]
@pditommaso I'm at a bit of a loss after hacking at this for hours- think I'm missing something obvious. Nothing I change seems to impact on the issue reported in https://github.com/nextflow-io/nextflow/issues/1577. 

Here's a simpler example:

```
./launch.sh kuberun ebi-gene-expression-group/rnatoy -v nextflow -c nextflow.config
```
(the toy edited to have a long process name)

The main issue appear to be the taskName, which seems to evade the modifications in this PR so far. It seemed to me that K8sTaskHandler.groovy would be the place to add the fix, but I haven't been able to make that work. 

In fact, it seems that some of the changes I make have no impact at all- e.g. if I change 'uuid' to 'uuid-blah' at  https://github.com/nextflow-io/nextflow/blob/415ac310c92ab14be7f1dbefea5534d8c97ff648/modules/nextflow/src/main/groovy/nextflow/k8s/K8sTaskHandler.groovy#L214 and do 'make compile', then run the workflow, the pods still end up without the 'blah' in the session ID. 

Could you point me in the right direction?

**Comment by pditommaso** [comment]
The problem is that using `kuberun` launch a pod with the prebuild container and therefore bypassing all your changes. Use `./launch.sh run <etc>` with the config highlighted [here](https://github.com/nextflow-io/nextflow/issues/1577#issuecomment-619859007). (sorry, I should have mentioned but I did't realise).

**Comment by pinin4fjords** [comment]
Aha! I knew I was missing something somewhere- thank you.

**Comment by pinin4fjords** [comment]
Sorry @pditommaso , still having an issue. On the plus side, I'm now seeing changes on the pod- good. On the negative I'm now getting " /bin/bash: .command.run: No such file or directory":

```
 > ./launch.sh run nextflow-io/hello -c nextflow.config

N E X T F L O W  ~  version 20.04.1-edge
Launching `nextflow-io/hello` [wise_coulomb] - revision: 1d43afc0ec [master]
WARN: The use of `echo` method is deprecated
executor >  k8s (4)
[2c/9e056a] process > sayHello (2) [100%] 1 of 1, failed: 1
WARN: Killing pending tasks (3)
Error executing process > 'sayHello (3)'

Caused by:
  Process `sayHello (3)` terminated for an unknown reason -- Likely it has been terminated by the external system

Command executed:

  echo 'Hello world!'

Command exit status:
  -

Command output:
  (empty)

Command wrapper:
  /bin/bash: .command.run: No such file or directory

Work dir:
  /Users/jmanning/projects/nextflow/work/21/572724fb5405944b4a4f9e95d9d4c5

Tip: when you have fixed the problem you can continue the execution adding the option `-resume` to the run command line
```

The pod describe looks like:

```
Name:         nf-21572724fb5405944b4a4f9e95d9d4c5
Namespace:    default
Priority:     0
Node:         minikube/192.168.64.6
Start Time:   Thu, 30 Apr 2020 12:52:23 +0100
Labels:       app=nextflow
              processName=sayHello
              runName=wise_coulomb
              sessionId=uuid-9ede7f66-ab92-4331-9df8-638ea37181e5
              taskName=sayHello_3
Annotations:  <none>
Status:       Failed
IP:           172.17.0.4
IPs:
  IP:  172.17.0.4
Containers:
  nf-21572724fb5405944b4a4f9e95d9d4c5:
    Container ID:  docker://9e2f24ab71b1313429408e44c49c114b4405c67e75ea8dddbad098517f565196
    Image:         nextflow/bash
    Image ID:      docker-pullable://nextflow/bash@sha256:28452ce0a624e8ce0332861c703df9f6c5bee470f789fb57c07d71e075baaf7f
    Port:          <none>
    Host Port:     <none>
    Command:
      /bin/bash
      -ue
      .command.run
    State:          Terminated
      Reason:       Error
      Exit Code:    1
      Started:      Thu, 30 Apr 2020 12:52:25 +0100
      Finished:     Thu, 30 Apr 2020 12:52:25 +0100
    Ready:          False
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /Users/jmanning/projects/nextflow/work/21/572724fb5405944b4a4f9e95d9d4c5 from vol-1 (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-m8xmj (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
Volumes:
  vol-1:
    Type:          HostPath (bare host directory volume)
    Path:          /Users/jmanning/projects/nextflow/work/21/572724fb5405944b4a4f9e95d9d4c5
    HostPathType:  
  default-token-m8xmj:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-m8xmj
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type    Reason     Age        From               Message
  ----    ------     ----       ----               -------
  Normal  Scheduled  <unknown>  default-scheduler  Successfully assigned default/nf-21572724fb5405944b4a4f9e95d9d4c5 to minikube
  Normal  Pulling    29s        kubelet, minikube  Pulling image "nextflow/bash"
  Normal  Pulled     27s        kubelet, minikube  Successfully pulled image "nextflow/bash"
  Normal  Created    27s        kubelet, minikube  Created container nf-21572724fb5405944b4a4f9e95d9d4c5
  Normal  Started    27s        kubelet, minikube  Started container nf-21572724fb5405944b4a4f9e95d9d4c5
```

**Comment by pinin4fjords** [comment]
That said, I can see that the taskName is being adjusted successfully, so the purpose of the PR is probably fulfilled if @pditommaso is happy with the changes.

**Comment by pinin4fjords** [comment]
Do you require any more changes @pditommaso ?

**Comment by pditommaso** [comment]
Hi, there are some tests failing. Having you tried running them locally?

**Comment by pinin4fjords** [comment]
@pditommaso - was just a random CI failure from when I updated from master (I checked and the error had nothing to do with any of my changes). I couldn't re-trigger the CI at the time, but just merged master again and the tests re-ran- all good!

**Comment by pditommaso** [comment]
Nice, I'll merge at my earliest convenience.

**Comment by pinin4fjords** [comment]
Thanks! Any clues as to why I was getting the '.command.run' error above? I guess all will be fine once this make it into a release, but would like to trial it in the meantime.

**Comment by pditommaso** [comment]
Ok, I've merged this manually because I've realised also annotations had the same issue.

**Pull request merged**